File: test-suite/golang/security/host_header_poisoning_buggy.go
package security

import (
	"fmt"
	"net/http"
	"net/url"
)

type resetMailer interface {
	Send(to string, html string) error
}

func passwordResetURL(r *http.Request, token string) string {
	host := r.Host
	return fmt.Sprintf("https://%s/reset?token=%s", host, url.QueryEscape(token))
}

func forwardedVerificationURL(r *http.Request, token string) string {
	host := r.Header.Get("X-Forwarded-Host")
	if host == "" {
		host = r.Host
	}
	return (&url.URL{
		Scheme:   "https",
		Host:     host,
		Path:     "/verify",
		RawQuery: "token=" + url.QueryEscape(token),
	}).String()
}

func sendResetEmail(r *http.Request, mailer resetMailer, token string) error {
	host := r.Header.Get("Host")
	resetLink := "https://" + host + "/account/reset?token=" + url.QueryEscape(token)
	return mailer.Send("user@example.com", `<a href="`+resetLink+`">Reset password</a>`)
}

func resetURLWithValidationTooLate(r *http.Request, token string) string {
	host := r.Host
	resetLink := "https://" + host + "/reset?token=" + url.QueryEscape(token)
	if !map[string]bool{"app.example.com": true}[host] {
		return ""
	}
	return resetLink
}

func canonicalURLFromRequest(r *http.Request, pathname string) string {
	origin := "https://" + r.Host
	return origin + pathname
}

File: test-suite/js/security/host-header-poisoning-buggy.ts
type HeaderRequest = {
  headers: Record<string, string | undefined>;
  get(name: string): string | undefined;
};

type WebRequest = {
  headers: {
    get(name: string): string | null;
  };
};

type Mailer = {
  send(message: { to: string; html: string }): Promise<void>;
};

declare function headers(): { get(name: string): string | null };

export function passwordResetUrl(req: HeaderRequest, token: string): string {
  const host = req.headers.host ?? "app.example.com";
  return `https://${host}/reset?token=${token}`;
}

export function forwardedVerificationUrl(request: WebRequest, token: string): string {
  const forwardedHost = request.headers.get("x-forwarded-host");
  const host = forwardedHost || request.headers.get("host") || "";
  const verifyUrl = new URL(`/verify?token=${token}`, `https://${host}`).toString();
  return verifyUrl;
}

export async function sendResetEmail(req: HeaderRequest, mailer: Mailer, token: string): Promise<void> {
  const host = req.get("host") ?? "app.example.com";
  const resetLink = `https://${host}/account/reset?token=${token}`;
  await mailer.send({
    to: "user@example.com",
    html: `<a href="${resetLink}">Reset password</a>`,
  });
}

export function nextHeadersCanonicalUrl(pathname: string): string {
  const origin = `https://${headers().get("host")}`;
  return new URL(pathname, origin).toString();
}
